File: segmentManager/src/main/java/com/clumob/segment/manager/SegmentManager.java
package com.clumob.segment.manager;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;

import com.clumob.log.AppLog;
import com.clumob.segment.controller.SegmentInfo;
import com.clumob.segment.controller.util.ParcelableUtil;
import com.clumob.segment.empty.EmptySegment;

/**
 * Created by prashant.rathore on 23/02/18.
 */

public class SegmentManager {

    public static final int SEGMENT_ID_EMPTY = Integer.MIN_VALUE;
    private final SegmentManager parentSegmentManager;
    private final SegmentCallbacks callbacks;
    private final int managerId;
    private final Handler mHandler = new Handler();

    private Segment segment;
    private SegmentViewHolder screenView;

    private Context context;
    private SegmentNavigation navigation;

    public SegmentManager(int managerId, Context context, SegmentCallbacks callbacks) {
        this(null, managerId, context, callbacks);
    }

    SegmentManager(SegmentManager parentSegmentManager, int managerId, Context context, SegmentCallbacks callbacks) {
        this.parentSegmentManager = parentSegmentManager;
        this.managerId = managerId;
        this.context = context;
        this.callbacks = callbacks;
        this.navigation = callbacks.createSegmentNavigation(this);
    }

    public SegmentCallbacks getCallbacks() {
        return callbacks;
    }

    public SegmentManager getRootSegmentManager() {
        if (this.parentSegmentManager == null) {
            return this;
        } else
            return this.parentSegmentManager.getRootSegmentManager();
    }

    public void onPreCreate(@Nullable Bundle savedInstanceState) {
        this.segment = createDefaultSegmentController(savedInstanceState);
    }

    public void onPostCreate() {
        attachSegment();
    }

    protected void attachSegment() {
        segment.onCreate();
        screenView = segment.createView(null);
        changeView(screenView.getView(), null);
        segment.bindView(screenView);
    }

    private Segment createDefaultSegmentController(Bundle savedInstanceState) {
        Segment segment;
        SegmentInfo segmentInfo = restoreSegment(savedInstanceState);
        if (segmentInfo == null) {
            segment = new EmptySegment(new SegmentInfo(SEGMENT_ID_EMPTY, null));
        } else {
            segment = callbacks.provideSegment(segmentInfo);
        }
        segment.attach( context, LayoutInflater.from(context));
        return segment;
    }

    protected SegmentInfo restoreSegment(Bundle savedInstanceState) {
        byte[] segmentInfoBytes = null;
        if (savedInstanceState == null) {
//            segmentInfoBytes = getIntent().getByteArrayExtra("SEGMENT_INFO");
        } else {
            segmentInfoBytes = savedInstanceState.getByteArray("SEGMENT_INFO");
        }
        SegmentInfo segmentInfo = null;
        try {
            if (segmentInfoBytes != null) {
                segmentInfo = ParcelableUtil.unmarshall(segmentInfoBytes, SegmentInfo.CREATOR);
            }
        } catch (Exception e) {
            AppLog.printStack(e);
        }
        return segmentInfo;
    }

    SegmentInfo changeSegment(SegmentInfo segmentInfo) {
        Segment newController = callbacks.provideSegment(segmentInfo);
        newController.attach(context, LayoutInflater.from(context));
        final Segment oldController = this.segment;
        final SegmentViewHolder newScreen = newController.createView(null);
        switch (oldController.currentState) {
            case CREATE:
                newController.onCreate();
                newController.bindView(newScreen);
                break;
            case START:
                newController.onCreate();
                newController.bindView(newScreen);
                newController.onStart();
                break;
            case RESUME:
                oldController.onPause();
                newController.onCreate();
                newController.bindView(newScreen);
                newController.onStart();
                newController.onResume();
                break;
            case PAUSE:
                newController.onCreate();
                newController.bindView(newScreen);
                newController.onStart();
                break;
            case STOP:
                newController.onCreate();
                newController.bindView(newScreen);
                break;
            case DESTROY:
                return segmentInfo;

        }
        changeView(newScreen.getView(), new Runnable() {
            public void run() {
                switch (oldController.currentState) {
                    case CREATE:
                        oldController.unBindView();
                        oldController.onDestroy();
                        break;
                    case PAUSE:
                    case RESUME:
                    case START:
                        oldController.onStop();
                        oldController.unBindView();
                        oldController.onDestroy();
                        break;
                    case STOP:
                        oldController.unBindView();
                        oldController.onDestroy();
                        break;
                    case DESTROY:
                        return;
                }
            }
        });
        this.segment = newController;
        this.screenView = newScreen;
        return oldController.getSegmentInfo();
    }

    protected void changeView(View newView, Runnable onCompleHandler) {
        callbacks.setSegmentView(newView);
        if (onCompleHandler != null) {
            mHandler.post(onCompleHandler);
        }
    }


    public void onStart() {
        segment.onStart();
    }

    public void onResume() {
        segment.onResume();
    }

    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        segment.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        segment.onActivityResult(requestCode, resultCode, data);
    }

    public void onPause() {
        segment.onPause();
    }

    public void onSaveInstanceState(Bundle outState) {
        try {
            SegmentInfo segmentInfo = segment.getSegmentInfo();
            byte[] marshall = ParcelableUtil.marshall(segmentInfo);
            outState.putByteArray("SEGMENT_INFO", marshall);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void onStop() {
        segment.onStop();
    }

    public boolean handleBackPressed() {
        return segment != null && (segment.handleBackPressed() || navigation.popBackStack());
    }

    public void onDestroy() {
        segment.onDestroy();
        this.segment.unBindView();
        this.screenView = null;
    }

    public SegmentNavigation getNavigation() {
        return navigation;
    }

    public interface SegmentCallbacks<SI extends SegmentInfo> {

        public Segment provideSegment(SI segmentInfo);

        public void setSegmentView(View view);

        SegmentNavigation createSegmentNavigation(SegmentManager segmentManager);

    }

}

File: segmentManager/src/main/java/com/clumob/segment/manager/SegmentManager.kt
package com.clumob.segment.manager

import android.content.Context
import android.content.Intent
import android.content.res.Configuration
import android.os.Bundle
import android.os.Handler
import android.view.LayoutInflater
import android.view.View
import com.clumob.log.AppLog
import com.clumob.segment.controller.SegmentInfo
import com.clumob.segment.controller.util.ParcelableUtil.marshall
import com.clumob.segment.controller.util.ParcelableUtil.unmarshall
import com.clumob.segment.empty.EmptySegment
import com.clumob.segment.manager.Segment.SegmentState

/**
 * Created by prashant.rathore on 23/02/18.
 */
class SegmentManager internal constructor(private val parentSegmentManager: SegmentManager?,
                                          managerId: Int,
                                          private val context: Context,
                                          val callbacks: SegmentCallbacks?,
                                          val layoutInflater: LayoutInflater) : SegmentLifecycle {

    private val mHandler = Handler()
    private var segment: Segment<*, *>? = null
    private var screenView: SegmentViewHolder<*, *>? = null
    val navigation: SegmentNavigation?

    constructor(managerId: Int, context: Context, callbacks: SegmentCallbacks?, layoutInflater: LayoutInflater)
            : this(null, managerId, context, callbacks, layoutInflater)

    val rootSegmentManager: SegmentManager
        get() = if (parentSegmentManager == null) {
            this
        } else parentSegmentManager.rootSegmentManager

    override fun onCreate(savedInstanceState: Bundle?) {
        segment = createDefaultSegmentController(savedInstanceState)
        segment!!.onCreate()
        attachSegment()
    }

    private fun attachSegment() {
        screenView = segment!!.createView(null)
        changeView(screenView!!.view, null)
        segment!!.bindView(screenView!!)
    }

    private fun createDefaultSegmentController(savedInstanceState: Bundle?): Segment<*, *> {
        val segment: Segment<*, *>
        val segmentInfo = restoreSegment(savedInstanceState)
        segment = segmentInfo?.let { createRestoreSegment(it) } ?: createEmptySegment()
        segment.attach(context, layoutInflater)
        return segment
    }

    private fun createEmptySegment(): Segment<*, *> {
        return EmptySegment(SegmentInfo(SEGMENT_ID_EMPTY, null))
    }

    protected fun restoreSegment(savedInstanceState: Bundle?): SegmentInfo? {
        var segmentInfoBytes: ByteArray? = null
        if (savedInstanceState == null) { //            segmentInfoBytes = getIntent().getByteArrayExtra("SEGMENT_INFO");
        } else {
            segmentInfoBytes = savedInstanceState.getByteArray("SEGMENT_INFO")
        }
        var segmentInfo: SegmentInfo? = null
        try {
            if (segmentInfoBytes != null) {
                segmentInfo = unmarshall(segmentInfoBytes, SegmentInfo.CREATOR)
            }
        } catch (e: Exception) {
            AppLog.printStack(e)
        }
        return segmentInfo
    }

    private fun createRestoreSegment(segmentInfo: SegmentInfo): Segment<*,*> {
        return if (segmentInfo.id == SEGMENT_ID_EMPTY) createEmptySegment() else callbacks!!.provideSegment(segmentInfo)!!
    }

    fun changeSegment(segmentInfo: SegmentInfo): SegmentInfo? {
        val newController = createRestoreSegment(segmentInfo)
        newController.attach(context, layoutInflater)
        val oldController = segment
        val newScreen = newController.createView(null)
        when (oldController!!.currentState) {
            SegmentState.CREATE -> {
                newController.onCreate()
                newController.bindView(newScreen)
            }
            SegmentState.START -> {
                newController.onCreate()
                newController.bindView(newScreen)
                newController.onStart()
            }
            SegmentState.RESUME -> {
                oldController.onPause()
                newController.onCreate()
                newController.bindView(newScreen)
                newController.onStart()
                newController.onResume()
            }
            SegmentState.PAUSE -> {
                newController.onCreate()
                newController.bindView(newScreen)
                newController.onStart()
            }
            SegmentState.STOP -> {
                newController.onCreate()
                newController.bindView(newScreen)
            }
            SegmentState.DESTROY -> return segmentInfo
        }
        changeView(newScreen.view, Runnable {
            when (oldController.currentState) {
                SegmentState.CREATE -> {
                    oldController.unBindView()
                    oldController.onDestroy()
                }
                SegmentState.PAUSE, SegmentState.RESUME, SegmentState.START -> {
                    oldController.onStop()
                    oldController.unBindView()
                    oldController.onDestroy()
                }
                SegmentState.STOP -> {
                    oldController.unBindView()
                    oldController.onDestroy()
                }
                SegmentState.DESTROY -> return@Runnable
            }
        })
        segment = newController
        screenView = newScreen
        return oldController.getSegmentInfo()
    }

    protected fun changeView(newView: View?, onCompleHandler: Runnable?) {
        callbacks!!.setSegmentView(newView)
        if (onCompleHandler != null) {
            mHandler.post(onCompleHandler)
        }
    }

    override fun onStart() {
        segment!!.onStart()
    }

    override fun onResume() {
        segment!!.onResume()
    }

    fun onConfigurationChanged(newConfig: Configuration?) {
        segment!!.onConfigurationChanged(newConfig)
    }

    fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        segment!!.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        segment!!.onActivityResult(requestCode, resultCode, data)
    }

    override fun onPause() {
        segment!!.onPause()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        try {
            val segmentInfo = segment!!.getSegmentInfo()
            val marshall = marshall(segmentInfo)
            outState.putByteArray("SEGMENT_INFO", marshall)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun onStop() {
        segment!!.onStop()
    }

    override fun handleBackPressed(): Boolean {
        return segment != null && (segment!!.handleBackPressed() || navigation != null && navigation.popBackStack())
    }

    override fun onDestroy() {
        segment!!.onDestroy()
        screenView = null
    }

    interface SegmentCallbacks {
        fun provideSegment(segmentInfo: SegmentInfo): Segment<*, *>?
        fun setSegmentView(view: View?)
        fun createSegmentNavigation(segmentManager: SegmentManager?): SegmentNavigation?
    }

    companion object {
        const val SEGMENT_ID_EMPTY = Int.MIN_VALUE
    }

    init {
        navigation = callbacks!!.createSegmentNavigation(this)
    }
}